### PR TITLE
[MPOM-244] upload SHA-512 only for source-release to staging repo

### PR DESCRIPTION
**Pull request by kwin**
The SHA-512 checksum should be attached similar to the source-release.

### CODE REVIEW

File: pom.xml
Comment (michael-o) [inline pom.xml:456]: Where did you get the `source-release-tar` classifier from?

Comment (kwin) [reply pom.xml:456]: My bad, I think all descriptors use the same classifier (https://github.com/apache/maven-resources/blob/e9eaa1baebb72795555eba015eac9b40f288e931/apache-source-release-assembly-descriptor/src/main/assemblies/source-release.xml#L22)

**Review by michael-o — CHANGES_REQUESTED** [review]
I consider this change non-sense and incomplete:
* ASF releases are solely distributed through dist.apache.org
* Generating checksums on the fly for remote repos is done by Maven Resolver
* Uploading checksums for a subset of items makes no sense

Note that Maven Resolver uses checksums to detect transport corruption only. Verification *must* happen with signatures.

**Comment by michael-o** [comment]
@hboutemy

**Comment by kwin** [comment]
> ASF releases are solely distributed through dist.apache.org

Not in all Apache projects, look at https://repo1.maven.org/maven2/org/apache/sling/org.apache.sling.api/2.9.0/ or https://repo1.maven.org/maven2/org/apache/felix/org.apache.felix.scr/2.1.8/. Maven Central has the source-release artifacts as well.

> Generating checksums on the fly for remote repos is done by Maven Resolver

Only MD5 and SHA1, SHA512 has been explicitly disabled for performance reasons (https://issues.apache.org/jira/browse/MRESOLVER-140)

> Uploading checksums for a subset of items makes no sense

ASF policy enforces that only for those releases. For the binaries it is never checked (because as you said, Maven will only use MD5/SHA1 to detect transport corruption)

**Comment by michael-o** [comment]
> 
> 
> > ASF releases are solely distributed through dist.apache.org
> 
> Not in all Apache projects, look at https://repo1.maven.org/maven2/org/apache/sling/org.apache.sling.api/2.9.0/ or https://repo1.maven.org/maven2/org/apache/felix/org.apache.felix.scr/2.1.8/. Maven Central has the source-release artifacts as well.

They must comply with: https://infra.apache.org/release-distribution.html. They can provide more if they like, but not less than that.

> > Generating checksums on the fly for remote repos is done by Maven Resolver
> 
> Only MD5 and SHA1. SHA512 has been explicitly disabled for performance reasons (https://issues.apache.org/jira/browse/MRESOLVER-140)

What holds you off to enable this?

> > Uploading checksums for a subset of items makes no sense
> 
> ASF policy enforces that only for those releases (https://infra.apache.org/release-distribution.html#sigs-and-sums). For the binaries it is never checked (because as you said, Maven will only use MD5/SHA1 to detect transport corruption) and ASF doesn't care about the binaries anyhow.

ASF never provides binary. The only official release is a source tarball/zip. Binaries are courtesy. So if you want to point someone to an Apache release, point to downloads.apache.org. That's it.

**Comment by kwin** [comment]
To be honest, it is a bit tough to discuss with you. Also language like "non-sense" does not really help. But let me try to respond to your complaints:

> They must comply with: https://infra.apache.org/release-distribution.html. They can provide more if they like, but not less than that.

Yes, even Apache Sling and Felix do that, but this is IMHO not relevant in the context of this PR.

> What holds you off to enable this?

This is only possible with command line options and not via pom.xml configuration. To distribute that among a team is much more effort than on relying on a plugin available from Maven Central which is even compatible with Apache Maven prior 3.8.1 (which is IMHO the first release shipping with Maven Resolver 1.6.x supporting SHA-512 on demand)

> ASF never provides binary. The only official release is a source tarball/zip. Binaries are courtesy. So if you want to point someone to an Apache release, point to downloads.apache.org. That's it.

Same as above. IMHO not really relevant in the context of this PR. In fact a lot of ASF projects provide binaries via Maven Central although ASF only mandates distribution of source archives. The whole point about this PR is being able to use the Staging Repo as the single source for dist and Maven Central (as that eases voting, validating a release candidate).

**Comment by michael-o** [comment]
Does [this](https://maven.apache.org/maven-release/maven-release-plugin/perform-mojo.html#arguments) work for you? I also believe that if you attach checksums Resolver will calculate checksums of checksums.

**Comment by kwin** [comment]
I decided against using maven-resolver-impl for calculating SHA-512 for the following reasons:
1. It requires Resolver 1.6+ which only ships with Maven 3.8.1
2. It will calculate SHA-512 for all artifacts instead only the source-release ones which is IMHO too much overhead and not really useful

Although the SHA-512 artifact being calculated by the checksum-maven-plugin will get another MD5/SHA1 checksum when being deployed that IMHO doesn't do any harm.

**Comment by kwin** [comment]
@hboutemy I would appreciate your input on this as well.

**Comment by hboutemy** [comment]
IIUC, in the end, you end-up about publishing sha512 checksum to Central **only for source-release**?
I must admit that this will ease people's filing Apache dist area, because copying from the build workspace is cumbersome, and I don't expect usual release manager thinks about it after doing 'mvn release:perform', then he has hard time to fill Apache dist area and recalculate the sha512 once the vote is done
I like this limited scope sha512 idea: @michael-o don't you struggle to get the sha512 when committing source-release to Apache dista area?

on the implementation side, this means that the current commit is big because there was POM reformatting (change of order of 2 plugins), but the real update is about adding `<attachChecksums>true</attachChecksums>` to checksum plugin for source-release: nice
question: is the reordering necessary?

**Comment by kwin** [comment]
The reordering is no longer necessary with m-gpg-p 3.0.1 I will revert

**Comment by hboutemy** [comment]
I understand why you need to exclude asc from checksum, if asc was produced before checksum
I don't understand why we need to ensure that asc was produced before checksum: if by chance it was not produced yet, let benefit from chance, isn't it?

**Comment by kwin** [comment]
Let me summarize the status quo (without PR):

1. `gpg-maven-plugin:sign` executed in phase `verify`, calculates ASC files for almost all attached files (except for MD5, SHA1 and ASC)
1. `checksum-maven-plugin:files` executed in phase `verify`, calculates the checksum based on filename (considering the extension) but does not attach it

With this PR the order is inverted:

1. `checksum-maven-plugin:artifacts` executed in phase `verify`, calculates the checksum based on classifier (disregarding the extension) and attaching it to project
1. `gpg-maven-plugin:sign` executed in phase `verify`, calculates ASC files for almost all attached files (except for MD5, SHA1, SHA-512 and ASC)

Without the inverted order the `checksum-maven-plugin` would calculate a checksum for the `ASC` for  `source-release` as well, which is certainly not desired.

**Comment by hboutemy** [comment]
ah, ok, I didn't get that attaching checksums required to switch from `files` goal to `artifacts`, and no exclusion based on extension exists yet...
thanks for the explanation
@michael-o do you get the idea of publishing `source-release` sha512 to remote repository to ease filling Apache distribution?

**Comment by hboutemy** [comment]
@kwin wouldn't `files` goal https://checksum-maven-plugin.nicoulaj.net/files-mojo.html be an option? It supports attach, and would not suffer from the ordering prerequisite (which is not something that is guaranteed in a phase

**Comment by kwin** [comment]
files goal does never attach with a classifier, therefore we can't use it

**Comment by hboutemy** [comment]
ok, then it's safer to bind the `checksum-maven-plugin:artifacts` goal to `post-integration-test` phase

**Comment by kwin** [comment]
I can do that but for clarity reason I would still reorder in the pin

**Review by michael-o — COMMENTED** [review]
I have checked the change. Two questions:
* Will Resolver generate a checksum of the checksum when the checkusm is attached?
* What will happen if Resolver is configured to generate SHA-512 sums?

**Comment by kwin** [comment]
I answered the first question already in https://github.com/apache/maven-apache-parent/pull/40#issuecomment-831703133:

> Although the SHA-512 artifact being calculated by the checksum-maven-plugin will get another MD5/SHA1 checksum when being deployed that IMHO doesn't do any harm.

For the 2nd question, I have not tried, but most probably resolver will just overwrite the existing file, for now this isn't an issue, because I don't see a reason why ASF projects should generate SHA-512 by Maven Resolver, it won't ever be enabled by default for performance reasons.

**Comment by michael-o** [comment]
> 
> 
> I answered the first question already in [#40 (comment)](https://github.com/apache/maven-apache-parent/pull/40#issuecomment-831703133):
> 
> > Although the SHA-512 artifact being calculated by the checksum-maven-plugin will get another MD5/SHA1 checksum when being deployed that IMHO doesn't do any harm.
> 
> For the 2nd question, I have not tried, but most probably resolver will just overwrite the existing file (with the same content), for now this isn't an issue, because I don't see a reason why ASF projects should generate SHA-512 by Maven Resolver, it won't ever be enabled by default for performance reasons.

Even if Resolver overwrites, Nexus Staging might reject this because the file is uploaded twice. You should check that.

**Comment by kwin** [comment]
I think this should be handled in the resolver in https://github.com/apache/maven-resolver/blob/17ea285d40e8cdd183d45b074f74aae6d9f3d3fc/maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java#L286.

**Comment by michael-o** [comment]
> 
> 
> I think this should be handled in the resolver in https://github.com/apache/maven-resolver/blob/17ea285d40e8cdd183d45b074f74aae6d9f3d3fc/maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java#L286.

Why? Then you need to add excludes and transport through layers...

**Comment by kwin** [comment]
I rather meant handled by throwing an exception if the checksum artifact name collides with one of the attached artifacts.

**Comment by hboutemy** [comment]
FYI, I tried `mvn -Papache-release deploy` with the sha512 attachments and got https://repository.apache.org/content/repositories/snapshots/org/apache/apache/24-SNAPSHOT/

**Comment by michael-o** [comment]
@hboutemy I wonder what will happen during a release: Maven Resolver uploads SHA-512, this change uploads it. Will Nexus staging reject it? This should be tested before merged.

**Comment by hboutemy** [comment]
the day Maven Resolver will be configured to upload sha512, we'll remove this hack (because, yes, it's a hack to ease our release process)
When do you expect Maven Resolver to publish sha512 in addition to sha1 and md5 in real world?

**Comment by michael-o** [comment]
> 
> 
> the day Maven Resolver will be configured to upload sha512, we'll remove this hack (because, yes, it's a hack to ease our release process)
> When do you expect Maven Resolver to publish sha512 in addition to sha1 and md5 in real world?

People can already configure it. I just want to avoid surprises for all Apache folks. E.g., Mark Thomas does this actively now.

**Comment by michael-o** [comment]
See here: https://repo1.maven.org/maven2/org/apache/tomcat/tomcat-catalina/10.0.6/

Produced by Resoler via Ant.

**Comment by hboutemy** [comment]
FYI, I added some instructions to calculate sha512 in dist: https://github.com/apache/maven-site/commit/29bc21e3dc9d2bb7ff61ce49cecf8a8600b9b5d1
and while at it, I recalculated checksums for all our latest releases and found 6 fingerprints not exactly using the expected format (fixed in revision 48099): this shows concrete example how people struggle to add the sha512 to dist area...

**Comment by hboutemy** [comment]
> People can already configure it. I just want to avoid surprises for all Apache folks. E.g., Mark Thomas does this actively now.

can you show me, please? I don't know how to do that

notice: I mean using Maven, because that's the Maven ASF parent POM that we're tweaking...

**Comment by michael-o** [comment]
> 
> 
> FYI, I added some instructions to calculate sha512 in dist: [apache/maven-site@29bc21e](https://github.com/apache/maven-site/commit/29bc21e3dc9d2bb7ff61ce49cecf8a8600b9b5d1)
> and while at it, I recalculated checksums for all our latest releases and found 6 fingerprints not exactly using the expected format (fixed in revision 48099): this shows concrete example how people struggle to add the sha512 to dist area...

I have done this massively in Core ITs: https://github.com/apache/maven-integration-testing/commit/672728a363c8b3503250589504efbfbea152e7e9

**Comment by michael-o** [comment]
> 
> 
> > People can already configure it. I just want to avoid surprises for all Apache folks. E.g., Mark Thomas does this actively now.
> 
> can you show me, please? I don't know how to do that
> 
> notice: I mean using Maven, because that's the Maven ASF parent POM that we're tweaking...

See https://maven.apache.org/resolver/configuration.html for `aether.checksums.algorithms=SHA-512`. Resolver should generate and update them. Please try in your setup.

**Comment by hboutemy** [comment]
oh nice, I didn't know that content!
I ran `mvn -Papache-release deploy -Daether.checksums.algorithms="SHA-512,SHA-1,MD5"` and got id 86 of https://repository.apache.org/content/repositories/snapshots/org/apache/apache/24-SNAPSHOT/
no problem for now

**Comment by hboutemy** [comment]
changed pom.xml version to 23 and ran the same `mvn -Papache-release deploy -Daether.checksums.algorithms="SHA-512,SHA-1,MD5"`
=> perfect staging repository: https://repository.apache.org/content/repositories/orgapacheapache-1020/org/apache/apache/23/

**Comment by michael-o** [comment]
> changed pom.xml version to 23 and ran the same `mvn -Papache-release deploy -Daether.checksums.algorithms="SHA-512,SHA-1,MD5"`
> => perfect staging repository: https://repository.apache.org/content/repositories/orgapacheapache-1020/org/apache/apache/23/

That is nice. One can manually delete the redundant files from staging.

**Review by michael-o — APPROVED** [review]

**Comment by hboutemy** [comment]
I did not test yet, but why don't we just add `<aether.checksums.algorithms>SHA-512,SHA-1,MD5</aether.checksums.algorithms>` property to the parent POM?
wouldn't it activate sha512 publishing by Maven for absolutely every artifact?

**Comment by kwin** [comment]
@hboutemy See my answer in https://github.com/apache/maven-apache-parent/pull/40#issuecomment-831703133

**Comment by hboutemy** [comment]
@kwin ok, thanks for the link
let's merge this

**Comment by michael-o** [comment]
@kwin Thanks for bearing with me. I am very picky when it affects all ASF committers.

**Comment by hboutemy** [comment]
reading your comment, it makes me feel that for a very long time, we're probably conflating:
- hash format for Maven repository format (= hash only, defined by Maven repository format and Maven code that reads/writes it)
- hash format for Apache distribution

**Comment by kwin** [comment]
ASF mandates this format: https://www.apache.org/info/verification.html

**Comment by khmarbaise** [comment]
The referenced page of [apache explicitly mentions](https://www.apache.org/info/verification.html) the usage of `shasum/md5sum/certUtil` tool which I have mentioned as well in [MNG-6784](https://issues.apache.org/jira/browse/MNG-6784). 

The format from my point of view is a consequence by using the appropriate tools... if not otherwise explicitly mentioned I would assume that the format is the one which can be consumed by the given tools. This is currently not the case.

**Comment by jira-importer** [comment]
Resolve #456

**Pull request merged**
